modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionService.scala

```diff
@@ -3,37 +3,65 @@ package org.tessellation.dag.l1.domain.transaction
+import org.tessellation.dag.l1.domain.transaction.ContextualTransactionValidator.NonContextualValidationError
+import org.tessellation.ext.cats.syntax.validated.validatedSyntax
+import org.tessellation.node.shared.domain.collateral.LatestBalances
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.node.shared.domain.transaction.TransactionValidator
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.snapshot.{Snapshot, SnapshotInfo, StateProof}
 import org.tessellation.schema.transaction.Transaction
 import org.tessellation.security.Hashed
 import org.tessellation.security.hash.Hash
 
+import fs2.Stream
+
+import ContextualTransactionValidator.ContextualTransactionValidationError
+
 trait TransactionService[F[_]] {
   def offer(transaction: Hashed[Transaction]): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]]
 }
 
 object TransactionService {
 
-  def make[F[_]: Async](
+  def make[
+    F[_]: Async,
+    P <: StateProof,
+    S <: Snapshot,
+    SI <: SnapshotInfo[P]
+  ](
     transactionStorage: TransactionStorage[F],
-    contextualTransactionValidator: ContextualTransactionValidator[F]
+    lastSnapshotStorage: LastSnapshotStorage[F, S, SI] with LatestBalances[F],
+    transactionValidator: TransactionValidator[F]
   ): TransactionService[F] = new TransactionService[F] {
 
     def offer(
       transaction: Hashed[Transaction]
     ): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]] =
-      contextualTransactionValidator.validate(transaction.signed).flatMap {
-        case Valid(_) =>
-          transactionStorage
-            .put(transaction)
-            .as(transaction.hash.asRight[NonEmptyList[ContextualTransactionValidationError]])
-        case Invalid(e) => e.toNonEmptyList.asLeft[Hash].pure[F]
+      lastSnapshotStorage.getCombinedStream.map {
+        case Some((s, si)) => (s.ordinal, si.balances.getOrElse(transaction.source, Balance.empty))
+        case None          => (SnapshotOrdinal.MinValue, Balance.empty)
+      }.changes.switchMap {
+        case (latestOrdinal, balance) =>
+          Stream.eval {
+            transactionValidator
+              .validate(transaction.signed)
+              .map(_.errorMap(NonContextualValidationError))
+              .flatMap {
+                case Valid(_) =>
+                  transactionStorage.put(transaction, latestOrdinal, balance)
+                case Invalid(e) =>
+                  e.toNonEmptyList.asLeft[Hash].leftWiden[NonEmptyList[ContextualTransactionValidationError]].pure[F]
+              }
+          }
+      }.head.compile.last.flatMap {
+        case Some(value) => value.pure[F]
+        case None =>
+          (new Exception(s"Unexpected state, stream should always emit the first snapshot"))
```

Comment:
I think this can happen right now, we don't check whether the node is ready on the endpoint (if the download finished). So it's kind of unexpected but not impossible under normal conditions.

Reply:
Also not sure if node turns to Ready only after download is done or if that can be earlier.

Reply:
Yeah. We should throw that eventually if node finished the Download but snapshot is not present.
But what if transaction comes if there is no download finished? All the validations which are based on ordinals may provide false negatives if we make a fallback to `SnapshotOrdinal.MinValue` and `Balance.empty`.

On the other hand better false negatives than false positives.